File: typescript/examples/poll-and-export-csv.ts
/**
 * Create a job, poll manually, export CSV. Requires GMF_API_KEY.
 */
import { writeFile } from "node:fs/promises";
import { Client, TERMINAL_STATUSES } from "../src/index.js";
import type { PlaceRow } from "../src/index.js";

const KEYWORD = "dentists in Austin TX";
const OUT = "leads.csv";

function toCsv(rows: PlaceRow[]): string {
  const fieldnames: string[] = [];
  const seen = new Set<string>();
  for (const row of rows) {
    for (const key of Object.keys(row)) {
      if (!seen.has(key)) {
        seen.add(key);
        fieldnames.push(key);
      }
    }
  }
  const escape = (v: string) =>
    /[",\n\r]/.test(v) ? `"${v.replace(/"/g, '""')}"` : v;
  const lines = [fieldnames.join(",")];
  for (const row of rows) {
    lines.push(fieldnames.map((k) => escape(String(row[k] ?? ""))).join(","));
  }
  return lines.join("\n") + "\n";
}

async function main() {
  const client = new Client();
  const created = await client.createJob(KEYWORD);
  const jobId = created.jobId;
  console.log(`jobId=${jobId}`);

  for (;;) {
    const job = await client.getJob(jobId);
    const status = (job.status ?? "").toLowerCase();
    console.log(`status=${status} rowCount=${job.rowCount}`);
    if (TERMINAL_STATUSES.has(status)) {
      if (status === "failed") {
        throw new Error(job.error || "job failed");
      }
      break;
    }
    await new Promise((r) => setTimeout(r, 2000));
  }

  const rows: PlaceRow[] = [];
  let cursor: string | null = "0";
  while (cursor !== null) {
    const page = await client.getResults(jobId, { limit: 100, cursor });
    if (page.rows) rows.push(...page.rows);
    cursor =
      page.nextCursor === undefined || page.nextCursor === null
        ? null
        : String(page.nextCursor);
  }

  await writeFile(OUT, toCsv(rows), "utf8");
  console.log(`Wrote ${rows.length} rows to ${OUT}`);
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: python/examples/poll_and_export_csv.py
#!/usr/bin/env python3
"""Create a job, poll manually, export CSV. Requires GMF_API_KEY."""

from __future__ import annotations

import csv
import time
from pathlib import Path

from gmaps_scraper import Client
from gmaps_scraper.models import TERMINAL_STATUSES

KEYWORD = "dentists in Austin TX"
OUT = Path("leads.csv")


def main() -> None:
    client = Client()
    created = client.create_job(KEYWORD)
    job_id = created["jobId"]
    print(f"jobId={job_id}")

    while True:
        job = client.get_job(job_id)
        status = (job.get("status") or "").lower()
        print(f"status={status} rowCount={job.get('rowCount')}")
        if status in TERMINAL_STATUSES:
            if status == "failed":
                raise SystemExit(job.get("error") or "job failed")
            break
        time.sleep(2)

    rows: list[dict] = []
    cursor: str | None = "0"
    while cursor is not None:
        page = client.get_results(job_id, limit=100, cursor=cursor)
        rows.extend(page.get("rows") or [])
        nxt = page.get("nextCursor")
        cursor = None if nxt is None else str(nxt)

    fieldnames: list[str] = []
    seen: set[str] = set()
    for row in rows:
        for key in row:
            if key not in seen:
                seen.add(key)
                fieldnames.append(key)

    with OUT.open("w", encoding="utf-8", newline="") as fh:
        writer = csv.DictWriter(fh, fieldnames=fieldnames or ["Name"])
        writer.writeheader()
        for row in rows:
            writer.writerow({k: row.get(k, "") for k in fieldnames})

    print(f"Wrote {len(rows)} rows to {OUT}")


if __name__ == "__main__":
    main()
